server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -2911,9 +2924,42 @@ public UserVm updateVirtualMachine(long id, String displayName, String group, Bo
             }
         }
 
+        if (!StringUtils.isEmpty(hostName)) {
+            vm.setHostName(hostName);
+            try {
+                updateDns(vm);
+            } catch (CloudRuntimeException e) {
+                throw new CloudRuntimeException(String.format("Failed to update hostname of VM %s to %s", vm.getInstanceName(), vm.getHostName()));
+            }
+        }
+
         return _vmDao.findById(id);
     }
 
+    private void updateDns(UserVmVO vm) {
+        List<NicVO> nicVOs = _nicDao.listByVmId(vm.getId());
+        for (NicVO nic : nicVOs) {
+            List<DomainRouterVO> routers = _routerDao.findByNetwork(nic.getNetworkId());
+            for (DomainRouterVO router : routers) {
+                Commands commands = new Commands(Command.OnError.Stop);
+                commandSetupHelper.createDhcpEntryCommand(router, vm, nic, false, commands);
+                try {
+                    if (!nwHelper.sendCommandsToRouter(router, commands)) {
```

Comment:
@Pearl1594 
what if router is not Running ?

Reply:
@weizhouapache  - Good point - I'll add a check. Should I fail the update in case there are any Routers shutoff ? Or merely skip it?

Reply:
@Pearl1594 skip it I think.